File: 0493-reverse-pairs/0493-reverse-pairs.cpp
#define vti vector<int>
#define ll long long
#define pb push_back

class Solution 
{
public:
    void merge(vector<int> &arr, int low, int mid, int high) 
    {
        vector<int> temp; 
        int left = low; 
        int right = mid + 1;
        while (left <= mid && right <= high) 
        {
            if (arr[left] <= arr[right]) 
                temp.push_back(arr[left++]);
            else 
                temp.push_back(arr[right++]);
        }
        while (left <= mid) 
            temp.push_back(arr[left++]);

        while (right <= high) 
            temp.push_back(arr[right++]);

        for (int i = low; i <= high; i++) 
            arr[i] = temp[i - low];
    }

    int countPairs(vector<int> &arr, int low, int mid, int high) 
    {
        int right = mid + 1;
        int cnt = 0;
        for (int i = low; i <= mid; i++) 
        {
            
            while (right <= high && (long long)arr[i] > 2*(long long)arr[right])
                right++;
            cnt += (right - (mid + 1));
        }
        return cnt;
    }

    int mergeSort(vector<int> &arr, int low, int high) 
    {
        int cnt = 0;
        if (low >= high) 
            return cnt;
        int mid = (low + high) / 2 ;
        cnt += mergeSort(arr, low, mid);
        cnt += mergeSort(arr, mid + 1, high);
        cnt += countPairs(arr, low, mid, high);
        merge(arr, low, mid, high);
        return cnt;
    }
    int reversePairs(vti& nums) 
    {
        int res = mergeSort(nums, 0, nums.size()-1);
        return res;
    }
};

File: 0493-reverse-pairs/0493-reverse-pairs.py
class Solution:
    def merge(self, arr: List[int], low:int, mid: int, high: int)->None:
        left = low
        right = mid+1
        temp = []
        
        while(left<=mid and right<=high):
            if(arr[left] <= arr[right]):
                temp.append(arr[left])
                left += 1
            else:
                temp.append(arr[right])
                right+= 1
            
        while(left<=mid):
            temp.append(arr[left])
            left += 1
        
        while(right <= high):
            temp.append(arr[right])
            right += 1
        
        for i in range(low, high+1):
            arr[i] = temp[i-low]
            
    def countPairs(self, arr: List[int], low: int, mid: int, high: int) -> int:
        count = 0
        right = mid+1
        for i in range(low, mid+1):
            while(right<=high and arr[i] > 2*arr[right]):
                right += 1
            count += (right - (mid+1))
        
        return count
            
    def mergeSort(self, arr: List[int], low:int, high: int)-> int:
        count = 0
        if(low>=high):
            return count
        
        mid = (low+high)//2
        count += self.mergeSort(arr, low, mid)
        count += self.mergeSort(arr, mid+1, high)
        count += self.countPairs(arr, low, mid, high)
        
        self.merge(arr, low, mid, high)
        return count
    def reversePairs(self, nums: List[int]) -> int:
        res = self.mergeSort(nums, 0, len(nums)-1)
        return res